internal/pkg/pluginmanager/manager.go

```diff
@@ -66,29 +80,63 @@ func CheckLocalPlugins(conf *configloader.Config) error {
 
 	for _, tool := range conf.Tools {
 		pluginFileName := configloader.GetPluginFileName(&tool)
+		pluginMD5FileName := configloader.GetPluginMD5FileName(&tool)
 		if _, err := os.Stat(filepath.Join(pluginDir, pluginFileName)); errors.Is(err, os.ErrNotExist) {
 			if err != nil {
 				return err
 			}
 			return fmt.Errorf("plugin %s doesn't exist", tool.Name)
 		}
-	}
+		if _, err := os.Stat(filepath.Join(pluginDir, pluginMD5FileName)); errors.Is(err, os.ErrNotExist) {
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf(".md5 file of plugin %s doesn't exist", tool.Name)
+		}
 
+		if err := checkPluginMismatch(pluginDir, pluginFileName, pluginMD5FileName, tool.Name); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
-func checkFileMD5(file string, dc *PbDownloadClient, pluginFileName string, version string) (bool, error) {
-	localmd5, err := LocalContentMD5(file)
+// checkPluginMismatch check if the plugins match with .md5
+func checkPluginMismatch(pluginDir, soFileName, md5FileName, tooName string) error {
+	isMD5Match, err := md5helper.ValidateFileMatchMD5(filepath.Join(pluginDir, soFileName), filepath.Join(pluginDir, md5FileName))
 	if err != nil {
-		return false, err
+		return err
 	}
-	remotemd5, err := dc.fetchContentMD5(pluginFileName, version)
-	if err != nil {
-		return false, err
+	if !isMD5Match {
+		return fmt.Errorf("plugin %s doesn't match with .md5", tooName)
 	}
+	return nil
+}
 
-	if localmd5 == remotemd5 {
-		return true, nil
+// redownloadPlugins re-download from remote
+func redownloadPlugins(dc *PbDownloadClient, pluginDir, pluginFileName, pluginMD5FileName, version string) error {
+	if err := os.Remove(filepath.Join(pluginDir, pluginFileName)); err != nil {
+		return err
+	}
+	if err := os.Remove(filepath.Join(pluginDir, pluginMD5FileName)); err != nil {
+		return err
 	}
-	return false, nil
+	// download .so file
+	if err := dc.download(pluginDir, pluginFileName, version); err != nil {
+		return err
+	}
+	// download .md5 file
+	if err := dc.download(pluginDir, pluginMD5FileName, version); err != nil {
+		return err
+	}
+	return nil
+}
+
+// DownloadDtmMD5 download remote dtm .md5 for compare with local .md5
+func DownloadDtmMD5(remoteMD5Dir, dtmFileName string) error {
```

Comment:
DtmMD5 looks funny. Either DownloadDTMsMD5/DownloadMD5ofDTM, or DownloadDtmMd5 (for acronyms, either all upper, or all only initial upper).

Reply:
updated